### ISSUE TITLE
Can this be used with async libraries?

### ISSUE THREAD

**Issue by Tosh0kan** [rsalmei/alive-progress#227]
I'm trying to figure out how to use Alive Progress with httpx, which is a requests libraries that can do multiple requests asynchronously. I can use just fine when it's just one connection or multiple connections done synchronously, but I'm struggling how to use this with multiple, asynchronous connections. Is it possible?

**Comment by Tosh0kan** [comment]
Yeah, I know. In fact, it was thanks to your snippet on issue #218 that led me to actually grasp the concept. What I'm having issues with is specifically how to access the multiple streams all at once. 

From my understanding, with an asynchronous requests library, the server will receive multiple requests at once, and will download multiple things in parallel through the `await asyncio.gather(*tasks)`. I have no idea how to get into that coroutine to use the `total = int(response.headers["Content-Length"])` on each of the urls of `tasks` them sum all of them use that as the total for alive-bar.

Also, saudações do Rio!

**Comment by rsalmei** [comment]
Ohh, vc é do Rio? Sou de SP, saudações 😉 

Anyway, there are better async examples here, just look for issues about asyncio or executors, especially the closed ones.

But no, asynchronous is NOTHING about parallel things, it is about concurrent things, which is very different. An async task will run on an async engine unhindered until it **has to stop** because of some pending IO operation, i.e. when it has to _await_ for external things. This makes it completely stop and another task, which can already make progress because its IO is ready, comes back and gets to run again. That's why you should never do CPU-bound operations on an async task, because it will make all other _concurrent_ tasks stop working.
Well, if you do still have any specific doubts, please provide an example code so I can help you make it work.

Inté!! 😜

**Comment by rsalmei** [comment]
Hey @Tosh0kan, any news here? Would you have some code so I can see what you are struggling with?

**Comment by Tosh0kan** [comment]
@rsalmei Oh, don't worry about it. The code I wanted to include this turned out to be unnecessary, since there was already a program that did what I wanted to do with it, and also because of IRL situations, I haven't been able to mess around with this lib. Sorry that I forgot to close the issue. Thank you for the care 😄

**Comment by rsalmei** [comment]
No problem 😉

**Comment by brutatoasta** [comment]
hey @rsalmei , im using python 3.11 and 3.1.4 of alive-progress.

It seems either asyncio taskgroups or asyncio.gather triggers  "UserWarning: Nested use of alive_progress is not yet supported." 
```python
from alive_progress import alive_bar
import asyncio

async def instance():
    async with alive_bar(title="self", bar =None, spinner ="waves", stats =None, elapsed = "{elapsed}", monitor=None) as bar:
    
        await asyncio.sleep(1)
        bar.title = "self2"
        await asyncio.sleep(1)
async def amain():
    # await asyncio.gather(instance(), instance())
    async with asyncio.TaskGroup() as tg:
        tg.create_task(instance())
        tg.create_task(instance())
if __name__ == "__main__":
    asyncio.run(amain())
```

Do you have any advice on how to proceed?

**Comment by rsalmei** [comment]
Yep, alive_bar spotless not be on the task itself, ever. The async loop will end up starting more than one task at the same time, and trigger a nested use...
The correct way would be to use alive_bar on a common center point like main() which starts tasks, then be notified somehow that tasks are being completed, like with asyncio.gather() or using queues.
I have some examples in closed tickets if you'd like to.

**Comment by brutatoasta** [comment]
Thanks for the speedy reply. I need multiple spinners for concurrent tasks, if you have examples that would work great. 

By the way, perhaps there could be an FAQ or Github Discussions page for async related questions since i see people have had related questions.

**Comment by rsalmei** [comment]
You're welcome. There's no support for multiple bars concurrently on screen. It is very complicated to manage a single cursor amongst several bars with different refresh rates, so I haven't tried it yet.
Good idea, I surely can open a discussion for that.

**Comment by allrobot** [comment]
> You're welcome. There's no support for multiple bars concurrently on screen. It is very complicated to manage a single cursor amongst several bars with different refresh rates, so I haven't tried it yet. Good idea, I surely can open a discussion for that.

Hello

1. Can you support real-time refreshing output for multiple progress bars? For example, the first line, second line, or x line display the progress bars and update them in real-time.

2. Your `alive-progress` is way cooler than `tqdm`, but it seems like it doesn't support asynchronous output? Could it be made to have APIs similar to `tqdm_asyncio.as_completed`, `asyncio.as_completed`, or `tqdm_asyncio.gather`?
https://github.com/tqdm/tqdm

>tqdm support asynchronous. I hope your project can do it
